Adds impersonation as a first class citizen

klizhentas opened the pull request:
```yaml
allow:
  impersonate:
    users: ['alice', 'bob']
    roles: ['*']
    where: 'contains(user.spec.traits["groups"], impersonate_role.traits)'
```

klizhentas committed:
Adds controls for impersonation requests.

Fixes #5352

```yaml
allow:
  impersonate:
    users: ['alice', 'bob']
    roles: ['*']
    where: 'contains(user.spec.traits["groups"], impersonate_role.traits)'
```

Adds "impersonator" to all X.509 and SSH client certs
issued using impersonation and does best effort to track
requests by impersonators in audit events.

Limits certs TTL to the impersonator's max session TTL.

Prevents impersonating users to recursively impersonate
other users.

Allows impersonating users to renew their own certificate,
for example to set route to cluster.

Adds missing token permission for editor role.

klizhentas commented:
@r0mant @fspmarshall can you folks please review? I'm adding more tests, but largely done with the implementation

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -1167,21 +1167,24 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
+	// this is typically done against a local user.
+	user, err := a.GetUser(req.Username, false)
+	if err != nil {
+		return nil, trace.AccessDenied("access denied")
```
Since you're replacing the original error with AccessDenied, I'd at least log `err` here so it's not lost.

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -1199,25 +1202,13 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
+	} else {
+		// Do not allow combining impersonation and access requests
+		if len(req.AccessRequests) > 0 {
+			return nil, trace.AccessDenied("access denied")
```
This will be pretty hard to troubleshoot, I'd consider logging this.

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -1257,16 +1248,41 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
+			}); err != nil {
+				log.WithError(err).Warn("Failed to emit local login failure event.")
+			}
+			return nil, trace.Wrap(err)
```
This will reveal a full `user %q has requested to generate ....` error message instead of generic `access denied` like in other places. Is this intentional?

klizhentas replied:
I thought that would not leak any specific information, because that's what client has supplied

r0mant commented on lib/services/impersonate.go:
```diff
@@ -0,0 +1,85 @@
+
+// getIdentifier returns identifier defined in a context
+func (ctx *impersonateContext) getIdentifier(fields []string) (interface{}, error) {
+	switch fields[0] {
```
Is `fields` guaranteed to have at least 1 element?

r0mant commented on lib/services/impersonate.go:
```diff
@@ -0,0 +1,85 @@
+// matchesImpersonateWhere returns true if Where rule matches.
+// Empty Where block always matches.
+func matchesImpersonateWhere(cond types.ImpersonateConditions, parser predicate.Parser) (bool, error) {
+	if cond.Where == "" {
```
Nit: `strings.TrimSpace(cond.Where)` or can it not consist of whitespaces only?

r0mant commented on lib/services/role.go:
```diff
@@ -505,6 +505,33 @@ func ApplyTraits(r Role, traits map[string][]string) Role {
+		}
+		outCond.Users = utils.Deduplicate(outCond.Users)
+		outCond.Roles = utils.Deduplicate(outCond.Roles)
+		r.SetImpersonateConditions(condition, outCond)
```
Don't you also need to copy Where to outCond? `outCond.Where = inCond.Where`. Otherwise seems like it might be lost here.

r0mant commented on lib/services/role.go:
```diff
@@ -1993,6 +2168,23 @@ const RoleSpecV3SchemaDefinitions = `
+				"type": "array",
+				"items": { "type": "string" }
+			  },
+		      "where": {
+                "type": "string"
```
Indentation seems off.

r0mant commented on lib/services/role.go:
```diff
@@ -1558,6 +1592,147 @@ func (set RoleSet) CheckAccessToKubernetes(namespace string, kube *KubernetesClu
+		if !roleMatch(impersonateRole.GetName()) {
+			return false, nil
+		}
+		ctx.impersonateRole = impersonateRole
```
Am I understanding correctly that this basically sets `impersonateRole` inside the `ctx` that is in turn used inside `whereParser` which is created in CheckImpersonate above but is being used right below?

If I'm correct and there's no easy way to make this more explicit (like maybe create `whereParser` here instead of in the caller CheckImpersonate), I'd at least add a comment explaining this, took me a minute to understand

fspmarshall reviewed (commented):
General thoughts:

- It should be possible to differentiate regular and impersonation certs in some manner (and possibly at least recover the root identity's name for audit log purposes).
- An impersonated identity probably should not be able to perform impersonation itself (i.e. impersonation should not be recursive).
- The lifetime of the impersonated cert should be limited by `min(root_cert_exp, now + impersonated_identity_session_ttl)`.  If folks want impersonation that escapes the expiry of the root identity's cert, that should be opt-in behavior (and if we do allow an escape mechanism, then impersonation *definitely* can't be recursive, or else you could just "ping-pong" between identities and your certs would never go invalid).
---


Additionally, I think we should have a rough idea of how we want this to interact with the upcoming locking feature before this gets merged.  IMO impersonated identities should be affected both by locks that affect the impersonated identity, and locks that affect the root identity.  If we think there is a chance that we will permit locks based on roles (likely I think), then things get extra tricky.  Consider this scenario:

```yaml
kind: user
metadata:
  name: alice
spec:
  roles: ['patriot']
---
kind: role
metadata:
  name: patriot
spec:
  allow:
    request:
      roles: ['spy']
  options:
    lockable: true # hypothetical
---
kind: role
metadata:
  name: spy
spec:
  allow:
    impersonate:
      users: ['bob']
      roles: ['enemy']
```
 
 If `alice` assumes role `spy` via the access request system, she will now be able to impersonate `bob` with role `enemy` (assuming thats a valid user+role combination).  Now say that a lock is applied which affects all users with role `spy`.  Since `alice`'s ability to act as `bob` is dependent upon role `spy`, she should probably lose the ability to act as `bob`.
 Since `alice` holding role `spy` isn't a static assignment, we can't just say "this lock applies to the root identity behind this cert".  We need to be able to say "this lock applies to the exact identity used to issue this impersonation cert".

fspmarshall commented on api/types/role.go:
```diff
@@ -592,7 +621,19 @@ func (r *RoleV3) CheckAndSetDefaults() error {
+		}
+	}
+	if r.Spec.Deny.Impersonate != nil {
+		if r.Spec.Deny.Impersonate.Where != "" {
+			return trace.BadParameter("'where' is not supported in deny.impersonate conditions")
+		}
```
Why not?  Seems like that would be useful.

klizhentas commented:
> It should be possible to differentiate regular and impersonation certs in some manner (and possibly at least recover the root >identity's name for audit log purposes).
> An impersonated identity probably should not be able to perform impersonation itself (i.e. impersonation should not be recursive).

Good points. I will bake in the impersonator in the cert, will add to the audit event and prevent recursive impersonation.

> The lifetime of the impersonated cert should be limited by min(root_cert_exp, now + impersonated_identity_session_ttl). If folks > want impersonation that escapes the expiry of the root identity's cert, that should be opt-in behavior (and if we do allow an >escape mechanism, then impersonation definitely can't be recursive, or else you could just "ping-pong" between identities and your certs would never go invalid).

Good point, I will restrict the TTL

klizhentas commented:
Re: locks

If we create a lock per identity, it should apply to certs of the impersonator. In your example, if we lock Alice by identity, then Alice's certs, including impersonated certs will be locked out, because her identity will be baked in the cert as impersonator.

If we create a lock per role, Alice's cert impersonating role that is not locked will not be locked as well, which I think is OK.

fspmarshall commented on lib/auth/auth_with_roles.go:
```diff
@@ -1167,21 +1167,25 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
+
+	// Extract the user and role set for whom the certificate will be generated.
+	// This should be safe since
+	// this is typically done against a local user.
```
> this is typically done against a local user.

Can we make this a hard rule?  Allowing impersonation of users who are managed by external identity providers feels like it would allow for some scenarios that are fairly difficult to reason about.  If we can disallow it, I think we should.  Something like:

```golang
if user.GetCreatedBy().Connector != nil {
    log.Warningf("User %v tried to issue a cert for externally managed user %v, this is not supported.", a.context.User.GetName(), req.Username)
    return trace.AccessDeneid("access denied")
}
```

klizhentas commented:
@fspmarshall take another look, I have:

* Prevented recursive impersonation
* Added impersonator in client SSH and TLS certs
* Limited TTL of impersonating users to the TTL of their cert
* Added impersonator to events and updated almost all audit events

fspmarshall commented on lib/auth/auth_with_roles.go:
```diff
@@ -1193,31 +1208,22 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
-		}); err != nil {
-			log.WithError(err).Warn("Failed to emit local login failure event.")
+	} else {
+		// Do not allow combining impersonation and access requests
+		if len(req.AccessRequests) > 0 {
+			log.WithError(err).Warningf("User %v tried to issue a cert for %v and added access requests. This is not supported.", a.context.User.GetName(), req.Username)
+			return nil, trace.AccessDenied("access denied")
 		}
```
Per my comment above, impersonated certs need to be able to perform normal reissues in order to set things like route-to-cluster directives.  To preserve this check's current behavior after that change, it should be moved to the outer scope and changed to something like this:

```golang
if a.context.Identity != nil && a.context.Identity.GetIdentity().Impersonator != "" && len(req.AccessRequests) > 0 {
    // ...
}
```

Also, note: When integrating this with the upcoming access review feature, we should prevent users from either reviewing their own requests under an assumed identity, or double-reviewing requests via assumed identities.

klizhentas commented:
@fspmarshall applied your comments on renewal of impersonated users, made sure they are not escaping, but allowed them to renew their own certs

fspmarshall commented on lib/auth/auth_with_roles.go:
```diff
@@ -1167,21 +1167,51 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
+	//
+	if a.context.Identity != nil && a.context.Identity.GetIdentity().Impersonator != "" {
+		if len(req.AccessRequests) > 0 {
+			return nil, trace.AccessDenied("access denied: impersonating user can not request new roles")
```
Nit: I think this should be "impersonated user" not "impersonating user".  The root identity is the one impersonating.  The identity being rejected here is being impersonated.

fspmarshall reviewed (approved):
@klizhentas LGMT.


(linking #5441 here to remind me that these are gonna need some care when merging)

russjones commented on lib/tlsca/parsegen.go:
```diff
@@ -218,6 +218,21 @@ func ParsePublicKeyDER(der []byte) (crypto.PublicKey, error) {
 
+// MarshalPrivateKeyPEM marshals RSA private key to PEM.
+// Current implementation returns error for any other type of private key.
+func MarshalPrivateKeyPEM(key interface{}) ([]byte, error) {
```
This is implemented in `utils.MarshalPrivateKey`.

russjones commented on lib/auth/auth_with_roles.go:
```diff
@@ -1257,14 +1278,45 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
+		req.Expires = a.authServer.GetClock().Now().Add(ttl)
+		if err != nil {
+			log.Warning(err)
+			err := trace.AccessDenied("user %q has requested to generate certs for %q.", a.context.User.GetName(), roles)
```
How will admins using Cloud debug this?

I don't have a good answer to that question at the moment either, but @alex-kovoy has mentioned that he has to fetch process logs to help customers debug SAML connectors. I imagine a similar situation will happen here.
